cli/src/diff_util.rs

```diff
@@ -2052,56 +2063,89 @@ pub fn show_diff_stats(
+        available_width.saturating_sub(max_path_width + diff_number_width + " ".len());
     let factor = if max_diffs < max_bar_length {
         1.0
     } else {
         max_bar_length as f64 / max_diffs as f64
     };
 
     for (stat, ui_path) in iter::zip(stats.entries(), &ui_paths) {
-        let bar_length = ((stat.added + stat.removed) as f64 * factor) as usize;
-        // Ensure that any fractional space after scaling is given to whichever
-        // of adds/removes is smaller.  This ensures we always show at least one
-        // tick even for small counts.
-        let (bar_added, bar_removed) = if stat.added < stat.removed {
-            let len = (stat.added as f64 * factor).ceil() as usize;
-            (len, bar_length - len)
-        } else {
-            let len = (stat.removed as f64 * factor).ceil() as usize;
-            (bar_length - len, len)
-        };
         // replace start of path with ellipsis if the path is too long
         let (path, path_width) = text_util::elide_start(ui_path, "...", max_path_width);
         let path_pad_width = max_path_width - path_width;
         write!(
             formatter,
-            "{path}{:path_pad_width$} | {:>number_width$}{}",
+            "{path}{:path_pad_width$} | ",
             "", // pad to max_path_width
-            stat.added + stat.removed,
-            if bar_added + bar_removed > 0 { " " } else { "" },
         )?;
-        write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
-        writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
+        if let Some((added, removed)) = stat.added_removed {
+            let bar_length = ((added + removed) as f64 * factor) as usize;
+            // Ensure that any fractional space after scaling is given to whichever
+            // of adds/removes is smaller.  This ensures we always show at least one
+            // tick even for small counts.
+            let (bar_added, bar_removed) = if added < removed {
+                let len = (added as f64 * factor).ceil() as usize;
+                (len, bar_length - len)
+            } else {
+                let len = (removed as f64 * factor).ceil() as usize;
+                (bar_length - len, len)
+            };
+            write!(
+                formatter,
+                "{:>diff_number_width$}{}",
+                added + removed,
+                if bar_added + bar_removed > 0 { " " } else { "" },
+            )?;
+            write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
+            writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
+        } else {
+            write!(formatter, "(binary)")?;
+            if stat.bytes_delta != 0 {
+                if stat.bytes_delta < 0 {
+                    write!(formatter.labeled("removed"), " {}", stat.bytes_delta)?;
+                } else if stat.bytes_delta > 0 {
+                    write!(formatter.labeled("added"), " +{}", stat.bytes_delta)?;
```

Comment:
nit: we could use `{:+}` instead, but I don't feel strongly.

Reply:
I think your idea is better, in particular because I can use it in both branches for symmetry.